src/link.rs

```diff
@@ -27,101 +30,142 @@ impl Link {
-            serde_json::from_slice(&bytes).map_err(|_| ArklibError::Parse)?;
-
-        Ok(Self { url, meta })
+        Ok(Self {
+            url,
+            prop: Properties {
+                title: user_prop.title,
+                desc: description,
+            },
+        })
     }
 
-    /// Write zipped file to path
-    pub async fn write_to_path<P: AsRef<Path>>(
-        &mut self,
+    pub async fn save<P: AsRef<Path>>(
+        &self,
         root: P,
-        path: P,
-        save_preview: bool,
+        with_preview: bool,
     ) -> Result<()> {
         let id = self.id()?;
-        store_meta::<Metadata, _>(root.as_ref(), id, &self.meta)?;
-
-        let mut link_file = File::create(path.as_ref().to_owned())?;
-        let file_data = self.url.as_str().as_bytes();
-        link_file.write(file_data)?;
-        if save_preview {
-            let preview_data = Link::get_preview(self.url.clone())
-                .await
-                .unwrap_or_default();
-            let image_data = preview_data
-                .fetch_image()
-                .await
-                .unwrap_or_default();
-            self.save_preview(root.as_ref(), image_data, id)
-                .await?;
+        let id_string = id.to_string();
+        let base_dir = root.as_ref().join(ARK_FOLDER);
+        let folder = base_dir
+            .join(LINK_STORAGE_FOLDER)
+            .join(&id_string);
+        let link_file = AtomicFile::new(&folder)?;
+        let tmp = link_file.make_temp()?;
+        (&tmp).write_all(self.url.as_str().as_bytes())?;
+        let current_link = link_file.load()?;
+        link_file.compare_and_swap(&current_link, tmp)?;
+
+        //User defined properties
+        let prop_folder = base_dir
+            .join(PROPERTIES_STORAGE_FOLDER)
+            .join(&id_string);
+        let prop_file = AtomicFile::new(prop_folder)?;
+        modify_json(&prop_file, |data: &mut Option<Properties>| {
+            let properties = self.prop.clone();
+            match data {
+                Some(data) => {
+                    // Hack currently overwrites
+                    *data = properties;
+                }
+                None => *data = Some(properties),
+            }
+        })?;
+
+        // Generated data
+        if let Ok(data) = self.get_preview().await {
+            let graph_folder = base_dir
+                .join(METADATA_STORAGE_FOLDER)
+                .join(&id_string);
+            let file = AtomicFile::new(graph_folder)?;
+            modify_json(&file, |file_data: &mut Option<OpenGraph>| {
+                let graph = data.clone();
+                println!("Trying to save: {with_preview} with {graph:?}");
+                match file_data {
+                    Some(file_data) => {
+                        // Hack currently overwrite
+                        *file_data = graph;
+                    }
+                    None => *file_data = Some(graph),
+                }
+            })?;
```

Comment:
Probably we should create `meta.rs` for handling generated metadata now. It would be similar to `prop.rs` but always overriding versions, because it's just a cache.